crates/ty_python_semantic/src/types/typed_dict.rs

```diff
@@ -246,14 +262,415 @@ impl<'db> TypedDictType<'db> {
+        }
+    }
+
+    /// Return the meta-type of this `TypedDict` type.
+    pub(super) fn to_meta_class(self, db: &'db dyn Db) -> ClassType<'db> {
+        // `TypedDict` instances are instances of `dict` at runtime, but its important that we
+        // understand a more specific meta type in order to correctly handle `__getitem__`.
+        match self {
+            TypedDictType::Class(defining_class) => defining_class,
+            TypedDictType::Synthesized(_) => KnownClass::TypedDictFallback
+                .try_to_class_literal(db)
+                .map(|class| class.default_specialization(db))
+                .unwrap_or_else(|| {
+                    KnownClass::Object
+                        .try_to_class_literal(db)
+                        .map(|class| class.default_specialization(db))
+                        .expect("object class must exist")
+                }),
+        }
+    }
+
+    pub(crate) fn normalized_impl(self, db: &'db dyn Db, visitor: &NormalizedVisitor<'db>) -> Self {
+        match self {
+            TypedDictType::Class(_) => {
+                let synthesized =
+                    SynthesizedTypedDictType::new(db, self.params(db), self.items(db));
+                TypedDictType::Synthesized(synthesized.normalized_impl(db, visitor))
+            }
+            TypedDictType::Synthesized(synthesized) => {
+                TypedDictType::Synthesized(synthesized.normalized_impl(db, visitor))
+            }
+        }
+    }
+
+    pub(crate) fn is_equivalent_to_impl(
+        self,
+        db: &'db dyn Db,
+        other: TypedDictType<'db>,
+        inferable: InferableTypeVars<'_, 'db>,
+        visitor: &IsEquivalentVisitor<'db>,
+    ) -> ConstraintSet<'db> {
+        if self.params(db) != other.params(db) {
+            return ConstraintSet::from(false);
+        }
+        // Compare the fields without requiring them to be in sorted order. Class-based `TypedDict`
+        // fields are not sorted. We do sort synthetic fields in `normalized_impl`, but there will
+        // soon be other sources of `SynthesizedTypedDictType` besides normalization.
+        if self.items(db).len() != other.items(db).len() {
+            return ConstraintSet::from(false);
+        }
+        let other_items = other.items(db);
+        let mut constraints = ConstraintSet::from(true);
+        for (name, field) in self.items(db) {
+            let Some(other_field) = other_items.get(name) else {
+                return ConstraintSet::from(false);
+            };
+            if field.kind != other_field.kind {
+                return ConstraintSet::from(false);
+            }
+            constraints.intersect(
+                db,
+                field.declared_ty.is_equivalent_to_impl(
+                    db,
+                    other_field.declared_ty,
+                    inferable,
+                    visitor,
+                ),
+            );
```

Comment:
I think you're missing an opportunity for an early return here:

```suggestion
            constraints.intersect(
                db,
                field.declared_ty.is_equivalent_to_impl(
                    db,
                    other_field.declared_ty,
                    inferable,
                    visitor,
                ),
            );
            if constraints.is_never_satisfied(db) {
                return constraints;
            }
```